### PR TITLE
NIFI-7234 Standardized on Avro 1.11.0

### PR DESCRIPTION
**Pull request by MikeThomsen**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by MikeThomsen** [comment]
It appears that Avro's default value behavior might have changed between 1.8 and 1.11. Gotta run down that issue. Verified that the build fails on both 1.11 and 1.10.

File: nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java
Comment (MikeThomsen) [inline nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java:329]: Changed this and our test schemas per this line in the Avro Specification:

> default: A default value for this field, only used when reading instances that lack the field for schema evolution purposes. The presence of a default value does not make the field optional at encoding time. Permitted values depend on the field's schema type, according to the table below. **Default values for union fields correspond to the first schema in the union**.

File: nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/test/resources/org/apache/nifi/avro/defaultArrayInRecords1.json
Comment (MikeThomsen) [inline nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/test/resources/org/apache/nifi/avro/defaultArrayInRecords1.json:19]: Not sure why the first one ever worked because the `0` literal is not an array. It broke with the update to 1.11

File: nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/test/resources/org/apache/nifi/avro/schema.json
Comment (MikeThomsen) [inline nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/test/resources/org/apache/nifi/avro/schema.json:7]: This appears to have changed with the parsing functionality in 1.9+. Under 1.10 and 1.11, the Avro API was not able to the first type and reference it in the second type until it was embedded in the second type as shown here.

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/test/java/org/apache/nifi/util/orc/TestNiFiOrcUtils.java
Comment (MikeThomsen) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/test/java/org/apache/nifi/util/orc/TestNiFiOrcUtils.java:385]: https://issues.apache.org/jira/browse/AVRO-1810

**Comment by MikeThomsen** [comment]
@exceptionfactory rebased and ready for review

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/pom.xml
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/pom.xml:74]: The version should be removed instead of commented out.
```suggestion
<!--            <version>1.8.1</version>-->
```
```suggestion
```

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/pom.xml
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/pom.xml:225]: Is it necessary to introduce the explicit dependency on `snappy-java`? Is it it no longer a transitive dependency of Avro?

File: nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/AbstractSiteToSiteReportingTask.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/AbstractSiteToSiteReportingTask.java:56]: This is a useful improvement, but is it required as part of this upgrade? Just looking to confirm the impact scope of the Avro upgrade.

Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/AbstractSiteToSiteReportingTask.java:56]: This is a useful improvement, but is it required as part of this upgrade? Just looking to confirm the impact scope of the Avro upgrade.

File: nifi-nar-bundles/nifi-sql-reporting-bundle/nifi-sql-reporting-tasks/pom.xml
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-sql-reporting-bundle/nifi-sql-reporting-tasks/pom.xml:85]: The version should be removed.
```suggestion
            <version>2.13.1</version>
```
```suggestion
```

Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-sql-reporting-bundle/nifi-sql-reporting-tasks/pom.xml:85]: The version should be removed.
```suggestion
            <version>2.13.1</version>
```
```suggestion
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONToSQL.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONToSQL.java:69]: Is this change required as part of the upgrade?

Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONToSQL.java:69]: Is this change required as part of the upgrade?

File: nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/pom.xml
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/pom.xml:140]: Is this dependency still required with the refactored classes that use Jackson 2?

Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/pom.xml:140]: Is this dependency still required with the refactored classes that use Jackson 2?

File: pom.xml
Comment (exceptionfactory) [inline pom.xml:277]: This comment should be removed.
```suggestion
<!--                <type>bundle</type>-->
```
```suggestion
```

Comment (exceptionfactory) [inline pom.xml:277]: This comment should be removed.
```suggestion
<!--                <type>bundle</type>-->
```
```suggestion
```

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/pom.xml
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/pom.xml:225]: Does not appear to be.

Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/pom.xml:225]: At least according to `mvn dependency:tree`

File: nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/AbstractSiteToSiteReportingTask.java
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/AbstractSiteToSiteReportingTask.java:56]: Those are Jackson 1.9.X references. 1.9.X was used in Avro 1.8.X and earlier. After that, it went to Jackson 2.X, so I think this is necessary.

File: nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/pom.xml
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/pom.xml:140]: Yes, until this is merged: https://github.com/apache/nifi/pull/5934

**Comment by MikeThomsen** [comment]
@exceptionfactory I think we're good to go now.

File: nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java
Comment (mattyb149) [reply nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java:329]: Does this affect fields for which the default value is `null`? I read in a couple of places (perhaps out of date) that in order to have a default value of `null`, then `null` has to be the first type in the union. Is that no longer the case?

**Comment by exceptionfactory** [comment]
Thanks for addressing the feedback @MikeThomsen, the current version looks good. I will defer to @mattyb149 for additional review, given the question on null handling for schema values.

File: nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java:329]: A nullable type should always be written as `[ "type_here"], "null" ]`. I'm not sure about a nullable type that also has a default value of null. That could be an edge case in the design of avro itself. What I do know is that Avro 1.9 and newer emphatically reject a default value of a real type if you do "null" first.

Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java:329]: Using this schema, I got the following validation error:

```
{
	"name": "Test",
	"type": "record",
	"fields": [
		{
			"name": "uuid",
			"type": [
				"string", "null"
			],
			"default": null
		}
	]
}
```
Screen Shot 2022-04-06 at 7 36 18 PM

AFAICT, my change is correct because that change was in the context of generating a nullable type, and that's how Avro's docs say to do a nullable type. If someone was able to get by using a bug in Avro's validation, I don't think that's something we should patch around.

Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java:329]: Reread the specification, and I'm not sure there even is a simple "correct" answer. What I do know is this: the sample schemas we had that worked under Avro 1.8, did not work under 1.10 and 1.11 because they were ordered "null", "type" and had a default value that was not null.

My gut says someone may end up having a bad day because of this dependency upgrade.

For example, I reverted this line locally, and this failed in `testCreateAvroSchemaPrimitiveTypes` in `TestAvroTypeUtil`:

```
fields.add(new RecordField("string", RecordFieldType.STRING.getDataType(), "hola",
                Collections.singleton("greeting")));
```

That string field is written turned into `[ "null", "string" ]` which Avro 1.9+ reject. Avro 1.9+ accept the opposite ordering.

This is the stacktrace; clearly it is an Avro bug that they fixed after 1.8 in their validation:

```
org.apache.avro.AvroTypeException: Invalid default for field string: "hola" not a ["null","string"]

	at org.apache.avro.Schema.validateDefault(Schema.java:1581)
	at org.apache.avro.Schema.access$500(Schema.java:92)
	at org.apache.avro.Schema$Field.<init>(Schema.java:552)
	at org.apache.avro.Schema$Field.<init>(Schema.java:588)
	at org.apache.nifi.avro.AvroTypeUtil.buildAvroField(AvroTypeUtil.java:135)
	at org.apache.nifi.avro.AvroTypeUtil.buildAvroSchema(AvroTypeUtil.java:121)
	at org.apache.nifi.avro.AvroTypeUtil.extractAvroSchema(AvroTypeUtil.java:101)
```

Comment (mattyb149) [reply nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java:329]: I think we are ok upgrading to something that actually enforces the spec, our inference stuff will not generate default values so the non-null type is supposed to come first. If the user has an invalid spec that got by previous versions' validation, it's something we can mention in Migration Guidance as something Avro is checking now.

**Review by exceptionfactory — APPROVED** [review]
Marking as approved, deferring to @mattyb149 for final confirmation and merging.

File: nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java
Comment (mattyb149) [reply nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java:329]: We should change inference to typ,null if it isn't in here already

Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java:329]: Poked around in the code for JSON and CSV readers and didn't see anything that was even creating obviously nullable fields.

**Comment by MikeThomsen** [comment]
@mattyb149 can we close this out or do you need more time to test with Hive and such?

**Comment by exceptionfactory** [comment]
It looks like there is a small conflict with the root `pom.xml` configuration, but otherwise this would be good to merge if @mattyb149 approves.

**Comment by mattyb149** [comment]
+1 LGTM, thanks for the upgrade! I fixed the conflict on merge to main

**Comment by MikeThomsen** [comment]
Thanks @mattyb149

**Pull request closed without merging**
